### PR TITLE
feat: add special keys to userEvent.type()

### PR DESCRIPTION
**Pull request by malwilley**
Hello! I hope it's okay if I submit a PR for this feature. It seems to be highly requested (https://github.com/testing-library/user-event/issues/181, https://github.com/testing-library/user-event/issues/172, https://github.com/testing-library/user-event/issues/31) and it would make this library much more useful for some of my own use cases. Was hoping this would be a smaller change, but the scope got away from me a little bit!

This adds support for some new key presses in `userEvent.type()`. I am following the same syntax that cypress uses in [their .type() command](https://docs.cypress.io/api/commands/type.html#Syntax).

Commands added:
- `{enter}`
- `{esc}`
- `{backspace}`
- `{shift}`
- `{ctrl}`
- `{alt}`
- `{meta}`

This does not yet track cursor position or selection range, so keys like `{leftarrow}` have not been implemented. This work could be done in a followup PR, but I wanted to keep the scope of this relatively small so I stopped here. 

Code notes:

- The code for the `type()` command was getting pretty unwieldy so I've broken it out into its own file.
- Testing that events were being called with the right properties was resulting in some hairy test code. To combat this I've added `toHaveBeenCalledWithEventAtIndex` which helps assert that the event handlers are being called with the correct events.

**Checklist**:

- [x] Documentation
- [x] Tests
- [ ] Typings (N/A)
- [x] Ready to be merged

### CODE REVIEW

**Comment by malwilley** [comment]
@kentcdodds @Gpx happy to see a flurry of activity on this repo!

I'd be happy to fix the conflicts on this PR if the changes are still wanted. Since this is such a big feature, however, I would not at all be offended if you preferred to tackle this in a different way. Let me know, and thank you for taking time to maintain this library!

**Comment by kentcdodds** [comment]
Hi! Thanks so much for this. Sorry for messing up the git history for you a bit 😬

I hadn't noticed this PR, but I'm definitely a fan. If you'd like to get it updated, I'll help get it merged. Thanks!

**Comment by malwilley** [comment]
@kentcdodds all cleaned up and ready for review!

Note that there are some prettier changes in `index.js` due to a previous PR getting merged without the new formatting requirements.

Also, code coverage is technically down in `index.js`, but I believe that's only because I reduced the number of lines in that file when I extracted the `type` function.

**Comment by kentcdodds** [comment]
Thanks for your work on this. Just want to let you know that I've seen this. I need more time to review it though. I hope to get back to you soon though.

File: src/type.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+  },
+  '{esc}': {
+    key: 'Escape',
+    code: 27,
+    skipPressEvent: true,
+  },
+  '{backspace}': {
+    key: 'Backspace',
+    code: 8,
+    inputType: 'deleteWordBackward',
```
Comment (klujanrosas) [inline src/type.js:23]: Should this be `deleteContentBackward`?

Comment (malwilley) [reply src/type.js:23]: Good catch! Yes it should

**Comment by kentcdodds** [comment]
Hi @malwilley,

I promise I'm still thinking about this PR. It's in my inbox and I think about it multiple times a day. I really feel like there's something we could do to improve the design and reduce code complexity, but I think I need to experiment with it myself a little bit. I may work on it a little bit today on my live stream at https://kcd.im/discord if you want to join in later.

**Comment by malwilley** [comment]
@kentcdodds No problem at all! I too think there is a lot of room for improvement -- I'm all ears if there is a more elegant solution to be found.

**Comment by kentcdodds** [comment]
Thanks for all your work on this! I've opened a new PR that I think you'll really like. Would love your feedback: #308

**Pull request closed without merging**